pkg/envoy/lds.go

```diff
@@ -1,41 +1,64 @@
 package envoy
 
 import (
-	"errors"
-
-	v1 "k8s.io/api/core/v1"
+	"fmt"
 
 	"github.com/openservicemesh/osm-health/pkg/common"
+	"github.com/openservicemesh/osm-health/pkg/osm"
 )
 
 // Verify interface compliance
-var _ common.Runnable = (*HasListener)(nil)
+var _ common.Runnable = (*HasListenerCheck)(nil)
 
-// HasListener implements common.Runnable
-type HasListener struct {
-	*v1.Pod
+// HasListenerCheck implements common.Runnable
+type HasListenerCheck struct {
+	ConfigGetter
+	osm.ControllerVersion
 }
 
 // Run implements common.Runnable
-func (l HasListener) Run() error {
-	envoyConfig, err := GetEnvoyConfig(l.Pod)
+func (l HasListenerCheck) Run() error {
+	if l.ConfigGetter == nil {
+		log.Error().Msg("Incorrectly initialized ConfigGetter")
+		return ErrIncorrectlyInitializedConfigGetter
+	}
+	envoyConfig, err := l.ConfigGetter.GetConfig()
 	if err != nil {
 		return err
 	}
+
 	if envoyConfig == nil {
-		return errors.New("envoy config is empty")
+		return ErrEnvoyConfigEmpty
 	}
+
+	expectedListenerName, exists := osm.ListenerNames[l.ControllerVersion]
+	if !exists {
+		return ErrOSMControllerVersionUnrecognized
+	}
+
+	if envoyConfig == nil || envoyConfig.Listeners.DynamicListeners == nil {
+		return ErrEnvoyConfigEmpty
+	}
+
+	actualListener := envoyConfig.Listeners.DynamicListeners[0]
+
+	if expectedListenerName != actualListener.Name {
+		log.Error().Msgf("must have listener with name %s but it is instead %s", expectedListenerName, actualListener.Name)
+		return ErrEnvoyListenerMissing
+	}
```

Comment:
Hey @draychev just for my understanding, in what situations does this case arise? If the pod has an envoy sidecar created through some other means (not by our mesh), but our mesh's own envoy sidecar in this pod failed to set up properly, then we have a case of mismatched envoy listener name?

Reply:
Good question -- Each source Envoy should have an outbound listener and each destination envoy should have an inbound listener.  take a look at this:  https://github.com/openservicemesh/osm-health/pull/25/files#diff-ec2be6376364a5c456575f111cccd27aea580f1b42d26bd4227a484ca1bb9c86R50-R51